docs/man/connection.3
.TH "connection" 3 "30 Apr 2004" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
connection \- Please note: classes serverrec and userrec both inherit from class connection. 

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <connection.h>\fP
.PP
Inherits \fBExtensible\fP.
.PP
Inherited by \fBserverrec\fP, and \fBuserrec\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBconnection\fP ()"
.br
.RI "\fIDefault constructor.\fP"
.ti -1c
.RI "bool \fBCreateListener\fP (char *\fBhost\fP, int p)"
.br
.RI "\fICreate a listening socket on 'host' using port number 'p'.\fP"
.ti -1c
.RI "bool \fBBeginLink\fP (char *targethost, int \fBport\fP, char *password, char *servername, int myport)"
.br
.RI "\fIBegin an outbound link to another ircd at targethost.\fP"
.ti -1c
.RI "bool \fBMeshCookie\fP (char *targethost, int \fBport\fP, long cookie, char *servername)"
.br
.RI "\fIBegin an outbound mesh link to another ircd on a network you are already an authenticated member of.\fP"
.ti -1c
.RI "void \fBTerminateLink\fP (char *targethost)"
.br
.RI "\fITerminate a link to 'targethost' by calling the \fBircd_connector::CloseConnection\fP method.\fP"
.ti -1c
.RI "bool \fBSendPacket\fP (char *message, const char *\fBhost\fP)"
.br
.RI "\fISend a message to a server by name, if the server is unavailable directly route the packet via another server If the server still cannot be reached after attempting to route the message remotely, returns false.\fP"
.ti -1c
.RI "bool \fBRecvPacket\fP (std::deque< std::string > &messages, char *\fBhost\fP)"
.br
.RI "\fIReturns the next available packet and returns true if data is available.\fP"
.ti -1c
.RI "\fBircd_connector\fP * \fBFindHost\fP (std::string \fBhost\fP)"
.br
.RI "\fIFind the \fBircd_connector\fP oject related to a certain servername given in 'host'.\fP"
.ti -1c
.RI "bool \fBAddIncoming\fP (int \fBfd\fP, char *targethost, int sourceport)"
.br
.RI "\fIAdd an incoming connection to the connection pool.\fP"
.ti -1c
.RI "long \fBGenKey\fP ()"
.br
.RI "\fIThis function is deprecated and may be removed in a later alpha/beta.\fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "int \fBfd\fP"
.br
.RI "\fIFile descriptor of the connection.\fP"
.ti -1c
.RI "char \fBhost\fP [256]"
.br
.RI "\fIHostname of connection.\fP"
.ti -1c
.RI "char \fBip\fP [32]"
.br
.RI "\fIIP of connection.\fP"
.ti -1c
.RI "char \fBinbuf\fP [MAXBUF]"
.br
.RI "\fIInbuf of connection.\fP"
.ti -1c
.RI "long \fBbytes_in\fP"
.br
.RI "\fIStats counter for bytes inbound.\fP"
.ti -1c
.RI "long \fBbytes_out\fP"
.br
.RI "\fIStats counter for bytes outbound.\fP"
.ti -1c
.RI "long \fBcmds_in\fP"
.br
.RI "\fIStats counter for commands inbound.\fP"
.ti -1c
.RI "long \fBcmds_out\fP"
.br
.RI "\fIStats counter for commands outbound.\fP"
.ti -1c
.RI "bool \fBhaspassed\fP"
.br
.RI "\fITrue if server/user has authenticated, false if otherwise.\fP"
.ti -1c
.RI "int \fBport\fP"
.br
.RI "\fIPort number For a userrec, this is the port they connected to the network on.\fP"
.ti -1c
.RI "int \fBregistered\fP"
.br
.RI "\fIUsed by userrec to indicate the registration status of the connection.\fP"
.ti -1c
.RI "short int \fBstate\fP"
.br
.RI "\fIReserved for future use.\fP"
.ti -1c
.RI "time_t \fBlastping\fP"
.br
.RI "\fITime the connection was last pinged.\fP"
.ti -1c
.RI "time_t \fBsignon\fP"
.br
.RI "\fITime the connection was created, set in the constructor.\fP"
.ti -1c
.RI "time_t \fBidle_lastmsg\fP"
.br
.RI "\fITime that the connection last sent data, used to calculate idle time.\fP"
.ti -1c
.RI "time_t \fBnping\fP"
.br
.RI "\fIUsed by PING checks with clients.\fP"
.ti -1c
.RI "char \fBinternal_addr\fP [MAXBUF]"
.br
.RI "\fIUnused, will be removed in a future alpha/beta.\fP"
.ti -1c
.RI "int \fBinternal_port\fP"
.br
.RI "\fIUnused, will be removed in a future alpha/beta.\fP"
.ti -1c
.RI "std::vector< \fBircd_connector\fP > \fBconnectors\fP"
.br
.RI "\fIWith a serverrec, this is a list of all established server connections.\fP"
.in -1c
.SH "Detailed Description"
.PP 
Please note: classes serverrec and userrec both inherit from class connection.Definition at line 161 of file connection.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "connection::connection ()"
.PP
Default constructor.
.SH "Member Function Documentation"
.PP 
.SS "bool connection::AddIncoming (int fd, char * targethost, int sourceport)"
.PP
Add an incoming connection to the connection pool.(reserved for core use)
.SS "bool connection::BeginLink (char * targethost, int port, char * password, char * servername, int myport)"
.PP
Begin an outbound link to another ircd at targethost.
.SS "bool connection::CreateListener (char * host, int p)"
.PP
Create a listening socket on 'host' using port number 'p'.
.SS "\fBircd_connector\fP* connection::FindHost (std::string host)"
.PP
Find the \fBircd_connector\fP oject related to a certain servername given in 'host'.
.SS "long connection::GenKey ()"
.PP
This function is deprecated and may be removed in a later alpha/beta.
.SS "bool connection::MeshCookie (char * targethost, int port, long cookie, char * servername)"
.PP
Begin an outbound mesh link to another ircd on a network you are already an authenticated member of.
.SS "bool connection::RecvPacket (std::deque< std::string > & messages, char * host)"
.PP
Returns the next available packet and returns true if data is available.Writes the servername the data came from to 'host'. If no data is available this function returns false. This function will automatically close broken links and reroute pathways, generating split messages on the network.
.SS "bool connection::SendPacket (char * message, const char * host)"
.PP
Send a message to a server by name, if the server is unavailable directly route the packet via another server If the server still cannot be reached after attempting to route the message remotely, returns false.
.SS "void connection::TerminateLink (char * targethost)"
.PP
Terminate a link to 'targethost' by calling the \fBircd_connector::CloseConnection\fP method.
.SH "Member Data Documentation"
.PP 
.SS "long connection::bytes_in"
.PP
Stats counter for bytes inbound.Definition at line 182 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "long connection::bytes_out"
.PP
Stats counter for bytes outbound.Definition at line 186 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "long connection::cmds_in"
.PP
Stats counter for commands inbound.Definition at line 190 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "long connection::cmds_out"
.PP
Stats counter for commands outbound.Definition at line 194 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "std::vector<\fBircd_connector\fP> connection::connectors"
.PP
With a serverrec, this is a list of all established server connections.With a userrec this is unused.Definition at line 241 of file connection.h.
.SS "int connection::fd"
.PP
File descriptor of the connection.Definition at line 166 of file connection.h.
.PP
Referenced by Server::SendTo(), serverrec::serverrec(), and userrec::userrec().
.SS "bool connection::haspassed"
.PP
True if server/user has authenticated, false if otherwise.Definition at line 198 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "char connection::host[256]"
.PP
Hostname of connection.Not used if this is a serverrecDefinition at line 170 of file connection.h.
.SS "time_t connection::idle_lastmsg"
.PP
Time that the connection last sent data, used to calculate idle time.Definition at line 224 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "char connection::inbuf[MAXBUF]"
.PP
Inbuf of connection.Only used for userrecDefinition at line 178 of file connection.h.
.SS "char connection::internal_addr[MAXBUF]"
.PP
Unused, will be removed in a future alpha/beta.Definition at line 232 of file connection.h.
.SS "int connection::internal_port"
.PP
Unused, will be removed in a future alpha/beta.Definition at line 236 of file connection.h.
.SS "char connection::ip[32]"
.PP
IP of connection.Reserved for future use.Definition at line 174 of file connection.h.
.SS "time_t connection::lastping"
.PP
Time the connection was last pinged.Definition at line 216 of file connection.h.
.PP
Referenced by serverrec::serverrec(), and userrec::userrec().
.SS "time_t connection::nping"
.PP
Used by PING checks with clients.Definition at line 228 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "int connection::port"
.PP
Port number For a userrec, this is the port they connected to the network on.For a serverrec this is the current listening port of the serverrec object.Definition at line 204 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "int connection::registered"
.PP
Used by userrec to indicate the registration status of the connection.Definition at line 208 of file connection.h.
.PP
Referenced by userrec::userrec().
.SS "time_t connection::signon"
.PP
Time the connection was created, set in the constructor.Definition at line 220 of file connection.h.
.PP
Referenced by serverrec::serverrec(), and userrec::userrec().
.SS "short int connection::state"
.PP
Reserved for future use.Definition at line 212 of file connection.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for InspIRCd from the source code.


docs/ChangeLog
1.0 Final

    * Optimized WriteServ - why use %s!%s@%s when we have the more optimal userrec::GetFullHost() ?
    * Added new event into m_spanningtree, event id send_metadata, which allows arbitary posting of metadata outside of the netburst, see line 529 of m_ssl_gnutls for an example of how to use it.
    * Added check that forbids configuration as root >:-)
    * Major updates to CBANs...like they're reasonably functional
    * Added support for TYPE_OTHER metadata, where void* target will be NULL, e.g. its not bound to a user or a nick
    * Fix to send 403 on invalid channel name
    * Moved Write() call so that it can send the NICK message properly and force the client to change
    * Allow ulines to still notice the channel regardless of +T
    * Created OnPostOper that can be gauranteed to be called *after* opering (by ALL modules) is done
    * "Oper of unknown type" -> "313 %s %s :is an IRC operator"
    * Auto-assume "+" on the start of opermode block if not included
    * Removed example commands that never actually existed
    * OpenSSL detection and cert generation in ./configure
    * Changed length of password field to 64, giving us a maximum password length of 63 plus null terminator
    * Fixed uninitialized ExtMode::list
    * Added some extra debug
    * Make m_services use InsertMode (are we done yet?)
    * Changed On005Numeric to use InsertMode (we left one out?)
    * Deleted the semicolon that had half the dev team wondering...
    * Incorporated Craig's friendlier SQL class (documentation to come later)
    * m_sqlauth: Fix small memory leaks in some cases (bad password, bad query, lost mysql server connection etc..)
    * Correctly initialize userrec::password (should fix a few m_sql issues)
    * Add m_conn_waitpong, require clients to reply with a correct PONG to a server PING before they can connect
    * Remove unused includes and rename classes from ModuleSSL to ModuleSSLGnuTLS so we can be consistent across multiple SSL modules
    * Added warnings to opers when session limits are exceeded
    * Improved IP handling. Now uses in_addr to store client ip, not char[16]!, Added global and local session limits
    * Problems removing +t due to wrong bit-value being used in comparison
    * Altered so that if running in -debug mode, will not erase its .so files from /tmp to allow a clean backtrace
    * Removed unneccessary strlcpy
    * Removed a load of unneeded includes (poll.h etc)
    * Added <ident:timeout> example so that people remember to set this when using m_ident.so
    * Default ident timeout to 1 sec if not defined (a bit low, but better than 0!)
    * Updated m_ident to use inspsocket, should be more reliable
    * Added missing modules (newer ones)
    * Removed excessive debug for something thats now fixed
    * Merged Andy Church's isnick() and isident() patches - faster, more efficient, and doesnt accidentally allow "-" as the first character of a nickname, Also adapted to iterate over a char* pointer rather than using unsigned int for the for() loops
    * Added m_opersha256 module created by Special
    * Add module for anti-autorejoin channel mode +J
    * Changes to INVEX and EXCEPTS 005 tokens
    * Changed +l to drop the mode if the limit is same as current., Optimized check on numeric chars
    * Remove last relic of last goto in inspircd (an unused label)
    * Fixed to not have ohv in CHANMODES type A,, Added proper MAXMODES output in 005 instead of hard coded '13'
    * One goto is now no gotos.
    * Change m_redirect, m_joinflood and m_messageflood to put their modes in the correct section and to use InsertMode()
    * Add InsertMode function to helperfuncs.(cpp|h) for easily adding modes to the CHANMODES section of the 005 numeric
    * Added patch from Andy Church re spaces in SETNAME
    * Fixed line endings when skipping ssl cert generation
    * Documented some functions that were missing comments
    * Properly check that existing key resides in conf/ and not current dir
    * Added PRIORITY_BEFORE and PRIORITY_AFTER (see src/modules/m_hostchange.cpp for how it works, function Prioritize())
    * Added -modupdate to regenerate modules makefile, Added SSL option, as well as a 'Build Certificate' section (If avaliable)
    * Adding the +e and +I modules m_banexception and m_inviteexception, also adding their shared utility header
    * Added trick to allow local server to unset -r only in one specific condition (the nickchange)
    * Added joinflood +j
    * Made bans check both real and fake host
    * Patched to allow FTOPIC from users (so that chanserv can update the settime and the topic still come from chanserv)
    * Added <link:hidden> to docs
    * Optional IP-Hiding for servers
    * Use libgnutls-config --libs to configure linker options
    * Make relative paths in the config file tags be relative to the insp config directory
    * Change 974 numeric to 490 to avoid collision with Insp's failed to load module error
    * Make m_sslmodes messages appear in the channel window
    * Updated with new docs for maxtargets
    * Updated 005 to show changeable MAXTARGETS
    * Allowed for configuration of maxtargets in <options:maxtargets>
    * Wrong nicklen is shown in the 005 numeric, fixed (noticed by ThaPrince)
    * Initial commit of m_sslmodes, supplying chanmode +z
    * Allow for <bind> tags with no or empty type, as well as 'clients' ones
    * Set vendor and static module flags...even though the booms from trying to unload it would be fun sad.gif
    * Make errors loading certificates DEFAULT level
    * Inital commit of the first SSL module \o/ \o/ \o/ Also Ommeh's first official svn commit \o/
    * Added MakeWildHost to userrec (*!*@dhost), Added ban stuff to messageflood
    * Fixes to server kick (i hope)
    * And yae, the compiler sayeth to brain, fix thine warning. and thou warning art fixeth.
    * Made '*' ban-specifier work

src/connection.cpp
/*       +------------------------------------+
 *       | Inspire Internet Relay Chat Daemon |
 *       +------------------------------------+
 *
 *  Inspire is copyright (C) 2002-2004 ChatSpike-Dev.
 *                       E-mail:
 *                <brain@chatspike.net>
 *           	  <Craig@chatspike.net>
 *     
 * Written by Craig Edwards, Craig McLure, and others.
 * This program is free but copyrighted software; see
 *            the file COPYING for details.
 *
 * ---------------------------------------------------
 */

#include <connection.h>
#include <unistd.h>
#include <fcntl.h>
#include <poll.h>
#include <sys/errno.h>
#include <sys/ioctl.h>
#include <sys/utsname.h>
#include <vector>
#include <string>
#include <deque>
#include <sstream>
#include "inspircd.h"
#include "modules.h"
#include "inspstring.h"
#include "helperfuncs.h"

using namespace std;


extern std::vector<Module*> modules;
extern std::vector<ircd_module*> factory;

std::deque<std::string> xsums;

extern int MODCOUNT;

extern time_t TIME;


/**
 * The InspIRCd mesh network is maintained by a tree of objects which reference *themselves*.
 * Every local server has an array of 32 *serverrecs, known as me[]. Each of these represents
 * a local listening port, and is not null if the user has opened a listening port on the server.
 * It is assumed nobody will ever want to open more than 32 listening server ports at any one
 * time (i mean come on, why would you want more, the ircd works fine with ONE).
 * Each me[] entry has multiple classes within it of type ircd_connector. These are stored in a vector
 * and each represents a server linked via this socket. If the connection was created outbound,
 * the connection is bound to the default ip address by using me[defaultRoute] (defaultRoute being
 * a global variable which indicates the default server to make connections on). If the connection
 * was created inbound, it is attached to the port the connection came in on. There may be as many
 * ircd_connector objects as needed in each me[] entry. Each ircd_connector implements the specifics
 * of an ircd connection in the mesh, however each ircd may have multiple ircd_connector connections
 * to it, to maintain the mesh link.
 */

char* xsumtable = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

// creates a random id for a line for detection of duplicate messages
std::string CreateSum()
{
	char sum[9];
	sum[0] = ':';
	sum[8] = '\0';
	for(int q = 1; q < 8; q++)
		sum[q] = xsumtable[rand()%52];
	return sum;
}

connection::connection()
{
	fd = 0;
}


bool connection::CreateListener(char* newhost, int p)
{
	sockaddr_in host_address;
	int flags;
	in_addr addy;
	int on = 0;
	struct linger linger = { 0 };
	
	this->port = p;
	
	fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
	if (fd <= 0)
	{
		return false;
	}

	setsockopt(fd,SOL_SOCKET,SO_REUSEADDR,(const char*)&on,sizeof(on));
	linger.l_onoff = 1;
	linger.l_linger = 1;
	setsockopt(fd,SOL_SOCKET,SO_LINGER,(const char*)&linger,sizeof(linger));
	
	// attempt to increase socket sendq and recvq as high as its possible
	// to get them on linux.
	int sendbuf = 32768;
	int recvbuf = 32768;
	setsockopt(fd,SOL_SOCKET,SO_SNDBUF,(const void *)&sendbuf,sizeof(sendbuf)); 
	setsockopt(fd,SOL_SOCKET,SO_RCVBUF,(const void *)&recvbuf,sizeof(sendbuf));

	memset((void*)&host_address, 0, sizeof(host_address));

	host_address.sin_family = AF_INET;

	if (!strcmp(newhost,""))
	{
		host_address.sin_addr.s_addr = htonl(INADDR_ANY);
	}
	else
	{
  		inet_aton(newhost,&addy);
		host_address.sin_addr = addy;
	}

	host_address.sin_port = htons(p);

	if (bind(fd,(sockaddr*)&host_address,sizeof(host_address))<0)
	{
		return false;
	}

	// make the socket non-blocking
	flags = fcntl(fd, F_GETFL, 0);
	fcntl(fd, F_SETFL, flags | O_NONBLOCK);

	this->port = p;

	listen(this->fd,5);

	return true;
}

char* ircd_connector::GetServerIP()
{
	return this->host;
}

int ircd_connector::GetServerPort()
{
	return this->port;
}

bool ircd_connector::SetHostAndPort(char* newhost, int newport)
{
	strncpy(this->host,newhost,160);
	this->port = newport;
	return true;
}

bool ircd_connector::SetHostAddress(char* newhost, int newport)
{
	strncpy(this->host,newhost,160);
	this->port = newport;
	memset((void*)&addr, 0, sizeof(addr));
	addr.sin_family = AF_INET;
	inet_aton(host,&addr.sin_addr);
	addr.sin_port = htons(port);
	return true;
}

void ircd_connector::SetServerPort(int p)
{
	this->port = p;
}

void ircd_connector::AddBuffer(std::string a)
{
	std::string b = "";
	for (int i = 0; i < a.length(); i++)
		if (a[i] != '\r')
			b = b + a[i];

	std::stringstream stream(ircdbuffer);
	stream << b;
	log(DEBUG,"AddBuffer: %s",b.c_str());
	ircdbuffer = stream.str();
}

bool ircd_connector::BufferIsComplete()
{
	for (int i = 0; i < ircdbuffer.length(); i++)
		if (ircdbuffer[i] == '\n')
			return true;
	return false;
}

void ircd_connector::ClearBuffer()
{
	ircdbuffer = "";
}

std::string ircd_connector::GetBuffer()
{
	// Fix by Brain 28th Apr 2005
	// seems my stringstream code isnt liked by linux
	// EVEN THOUGH IT IS CORRECT! Fixed by using a different
	// (SLOWER) algorithm...
        char* line = (char*)ircdbuffer.c_str();
        std::string ret = "";
        while ((*line != '\n') && (strlen(line)))
        {
                ret = ret + *line;
                line++;
        }
        if ((*line == '\n') || (*line == '\r'))
                line++;
        ircdbuffer = line;
        return ret;
}

bool ircd_connector::MakeOutboundConnection(char* newhost, int newport)
{
	log(DEBUG,"MakeOutboundConnection: Original param: %s",newhost);
	ClearBuffer();
	hostent* hoste = gethostbyname(newhost);
	if (!hoste)
	{
		log(DEBUG,"MakeOutboundConnection: gethostbyname was NULL, setting %s",newhost);
		this->SetHostAddress(newhost,newport);
		SetHostAndPort(newhost,newport);
	}
	else
	{
		struct in_addr* ia = (in_addr*)hoste->h_addr;
		log(DEBUG,"MakeOutboundConnection: gethostbyname was valid, setting %s",inet_ntoa(*ia));
		this->SetHostAddress(inet_ntoa(*ia),newport);
		SetHostAndPort(inet_ntoa(*ia),newport);
	}

	this->fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
	if (this->fd >= 0)
	{
		if(connect(this->fd, (sockaddr*)&this->addr,sizeof(this->addr)))
		{
			WriteOpers("connect() failed for %s",host);
			RemoveServer(this->servername.c_str());
			return false;
		}
		int flags = fcntl(this->fd, F_GETFL, 0);
		fcntl(this->fd, F_SETFL, flags | O_NONBLOCK);
		int sendbuf = 32768;
		int recvbuf = 32768;
		setsockopt(this->fd,SOL_SOCKET,SO_SNDBUF,(const void *)&sendbuf,sizeof(sendbuf)); 
		setsockopt(this->fd,SOL_SOCKET,SO_RCVBUF,(const void *)&recvbuf,sizeof(sendbuf));
		return true;
	}
	else
	{
		WriteOpers("socket() failed!");
		RemoveServer(this->servername.c_str());
	}

	return false;
}


bool connection::BeginLink(char* targethost, int newport, char* password, char* servername, int myport)
{
	char connect[MAXBUF];
	
	ircd_connector connector;
	ircd_connector *cn = this->FindHost(servername);


	if (cn)
	{
		WriteOpers("CONNECT aborted: Server %s already exists",servername);
		return false;
	}

	
	if (this->fd)
	{
		if (connector.MakeOutboundConnection(targethost,newport))
		{
			// targethost has been turned into an ip...
			// we dont want this as the server name.
			connector.SetServerName(servername);
			snprintf(connect,MAXBUF,"S %s %s %lu %lu :%s",getservername().c_str(),password,(unsigned long)myport,(unsigned long)GetRevision(),getserverdesc().c_str());
			connector.SetState(STATE_NOAUTH_OUTBOUND);
			connector.SetHostAndPort(targethost, newport);
			this->connectors.push_back(connector);
			return this->SendPacket(connect, servername);
		}
		else
		{
			connector.SetState(STATE_DISCONNECTED);
			WriteOpers("Could not create outbound connection to %s:%d",targethost,newport);
		}
	}
	return false;
}

void ircd_connector::SetVersionString(std::string newversion)
{
	log(DEBUG,"Set version of %s to %s",this->servername.c_str(),newversion.c_str());
	this->version = newversion;
}

std::string ircd_connector::GetVersionString()
{
	return this->version;
}

bool connection::MeshCookie(char* targethost, int newport, unsigned long cookie, char* servername)
{
	char connect[MAXBUF];
	
	ircd_connector connector;
	
	WriteOpers("Establishing meshed link to %s:%d",servername,newport);

	if (this->fd)
	{
		if (connector.MakeOutboundConnection(targethost,newport))
		{
			// targethost has been turned into an ip...
			// we dont want this as the server name.
			connector.SetServerName(servername);
			snprintf(connect,MAXBUF,"- %lu %s :%s",cookie,getservername().c_str(),getserverdesc().c_str());
			connector.SetState(STATE_NOAUTH_OUTBOUND);
			connector.SetHostAndPort(targethost, newport);
			connector.SetState(STATE_CONNECTED);
			this->connectors.push_back(connector);
			return this->SendPacket(connect, servername);
		}
		else
		{
			connector.SetState(STATE_DISCONNECTED);
			WriteOpers("Could not create outbound connection to %s:%d",targethost,newport);
		}
	}
	return false;
}

bool connection::AddIncoming(int newfd, char* targethost, int sourceport)
{
	ircd_connector connector;
	
	// targethost has been turned into an ip...
	// we dont want this as the server name.
	connector.SetServerName(targethost);
	connector.SetDescriptor(newfd);
	connector.SetState(STATE_NOAUTH_INBOUND);
	int flags = fcntl(newfd, F_GETFL, 0);
	fcntl(newfd, F_SETFL, flags | O_NONBLOCK);
	int sendbuf = 32768;
	int recvbuf = 32768;
	setsockopt(newfd,SOL_SOCKET,SO_SNDBUF,(const void *)&sendbuf,sizeof(sendbuf)); 
	setsockopt(newfd,SOL_SOCKET,SO_RCVBUF,(const void *)&recvbuf,sizeof(sendbuf));
	connector.SetHostAndPort(targethost, sourceport);
	connector.SetState(STATE_NOAUTH_INBOUND);
	log(DEBUG,"connection::AddIncoming() Added connection: %s:%d",targethost,sourceport);
	this->connectors.push_back(connector);
	return true;
}

void connection::TerminateLink(char* targethost)
{
	// this locates the targethost in the connection::connectors vector of the class,
 	// and terminates it by sending it an SQUIT token and closing its descriptor.
	// TerminateLink with a null string causes a terminate of ALL links
}


// Returns a pointer to the connector for 'host'
ircd_connector* connection::FindHost(std::string findhost)
{
	for (int i = 0; i < this->connectors.size(); i++)
	{
		if (this->connectors[i].GetServerName() == findhost)
		{
			return &this->connectors[i];
		}
	}
	return NULL;
}

std::string ircd_connector::GetServerName()
{
	return this->servername;
}

std::string ircd_connector::GetDescription()
{
	return this->description;
}

void ircd_connector::SetServerName(std::string serv)
{
	this->servername = serv;
}

void ircd_connector::SetDescription(std::string desc)
{
	this->description = desc;
}


int ircd_connector::GetDescriptor()
{
	return this->fd;
}

int ircd_connector::GetState()
{
	return this->state;
}


void ircd_connector::SetState(int newstate)
{
	this->state = newstate;
	if (state == STATE_DISCONNECTED)
	{
		NetSendMyRoutingTable();
	}
}

void ircd_connector::CloseConnection()
{
	int flags = fcntl(this->fd, F_GETFL, 0);
	fcntl(this->fd, F_SETFL, flags ^ O_NONBLOCK);
	close(this->fd);
	flags = fcntl(this->fd, F_GETFL, 0);
	fcntl(this->fd, F_SETFL, flags | O_NONBLOCK);
}

void ircd_connector::SetDescriptor(int newfd)
{
	this->fd = newfd;
}

bool connection::SendPacket(char *message, const char* sendhost)
{
	if ((!message) || (!sendhost))
		return true;

	ircd_connector* cn = this->FindHost(sendhost);
	
	if (!strchr(message,'\n'))
	{
		strlcat(message,"\n",MAXBUF);
	}

	if (cn)
	{
		log(DEBUG,"main: Connection::SendPacket() sent '%s' to %s",message,cn->GetServerName().c_str());
		
		if (cn->GetState() == STATE_DISCONNECTED)
		{
			log(DEBUG,"\n\n\n\nMain route to %s is down, seeking alternative\n\n\n\n",sendhost);
			// fix: can only route one hop to avoid a loop
			if (strncmp(message,"R ",2))
			{
				log(DEBUG,"Not a double reroute");
				// this route is down, we must re-route the packet through an available point in the mesh.
				for (int k = 0; k < this->connectors.size(); k++)
				{
					log(DEBUG,"Check connector %d: %s",k,this->connectors[k].GetServerName().c_str());
					// search for another point in the mesh which can 'reach' where we want to go
					for (int m = 0; m < this->connectors[k].routes.size(); m++)
					{
						log(DEBUG,"Check connector %d: %s route %s",k,this->connectors[k].GetServerName().c_str(),this->connectors[k].routes[m].c_str());
						if (!strcasecmp(this->connectors[k].routes[m].c_str(),sendhost))
						{
							log(DEBUG,"Found alternative route for packet: %s",this->connectors[k].GetServerName().c_str());
							char buffer[MAXBUF];
							snprintf(buffer,MAXBUF,"R %s %s",sendhost,message);
							this->SendPacket(buffer,this->connectors[k].GetServerName().c_str());
							return true;
						}
					}
				}
			}
			char buffer[MAXBUF];
			snprintf(buffer,MAXBUF,"& %s",sendhost);
			NetSendToAllExcept(sendhost,buffer);
			log(DEBUG,"\n\nThere are no routes to %s, we're gonna boot the server off!\n\n",sendhost);
			DoSplit(sendhost);
			return false;
		}

		// returns false if the packet could not be sent (e.g. target host down)
		if (send(cn->GetDescriptor(),message,strlen(message),0)<0)
		{
			log(DEBUG,"send() failed for Connection::SendPacket(): %s",strerror(errno));
			log(DEBUG,"Disabling connector: %s",cn->GetServerName().c_str());
			cn->CloseConnection();
			cn->SetState(STATE_DISCONNECTED);
			// retry the packet along a new route so either arrival OR failure are gauranteed (bugfix)
			return this->SendPacket(message,sendhost);
		}
		return true;
	}
}

bool already_have_sum(std::string sum)
{
	for (int i = 0; i < xsums.size(); i++)
	{
		if (xsums[i] == sum)
		{
			return true;
		}
	}
	if (xsums.size() >= 128)
	{
		xsums.pop_front();
	}
	xsums.push_back(sum);
	return false;
}

// receives a packet from any where there is data waiting, first come, first served
// fills the message and host values with the host where the data came from.

bool connection::RecvPacket(std::deque<std::string> &messages, char* recvhost,std::deque<std::string> &sums)
{
	char data[65536];
	memset(data, 0, 65536);
	for (int i = 0; i < this->connectors.size(); i++)
	{
		if (this->connectors[i].GetState() != STATE_DISCONNECTED)
		{
			// returns false if the packet could not be sent (e.g. target host down)
			int rcvsize = 0;

			// check if theres any data on this socket
			// if not, continue onwards to the next.
			pollfd polls;
			polls.fd = this->connectors[i].GetDescriptor();
			polls.events = POLLIN;
			int ret = poll(&polls,1,1);
			if (ret <= 0) continue;

			rcvsize = recv(this->connectors[i].GetDescriptor(),data,65000,0);
			data[rcvsize] = '\0';
			if (rcvsize == -1)
			{
				if (errno != EAGAIN)
				{
					log(DEBUG,"recv() failed for Connection::RecvPacket(): %s",strerror(errno));
					log(DEBUG,"Disabling connector: %s",this->connectors[i].GetServerName().c_str());
					this->connectors[i].CloseConnection();
					this->connectors[i].SetState(STATE_DISCONNECTED);
				}
			}
			int pushed = 0;
			if (rcvsize > 0)
			{
				this->connectors[i].AddBuffer(data);
				if (this->connectors[i].BufferIsComplete())
				{
					while (this->connectors[i].BufferIsComplete())
					{
						std::string text = this->connectors[i].GetBuffer();
						if (text != "")
						{
							if ((text[0] == ':') && (text.find(" ") != std::string::npos))
							{
								std::string orig = text;
								log(DEBUG,"Original: %s",text.c_str());
								std::string sum = text.substr(1,text.find(" ")-1);
								text = text.substr(text.find(" ")+1,text.length());
								std::string possible_token = text.substr(1,text.find(" ")-1);
								if (possible_token.length() > 1)
								{
									sums.push_back("*");
									text = orig;
									log(DEBUG,"Non-mesh, non-tokenized string passed up the chain");
								}
								else
								{
									log(DEBUG,"Packet sum: '%s'",sum.c_str());
									if ((already_have_sum(sum)) && (sum != "*"))
									{
										// we don't accept dupes
										log(DEBUG,"Duplicate packet sum %s from server %s dropped",sum.c_str(),this->connectors[i].GetServerName().c_str());
										continue;
									}
									sums.push_back(sum.c_str());
								}
							}
							else sums.push_back("*");
							messages.push_back(text.c_str());
							strlcpy(recvhost,this->connectors[i].GetServerName().c_str(),160);
							log(DEBUG,"Connection::RecvPacket() %d:%s->%s",pushed++,recvhost,text.c_str()); 
						}
					}
					return true;
				}
			}
		}
	}
	// nothing new yet -- message and host will be undefined
	return false;
}

